orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/tasks/StageStatusPreconditionTask.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2019 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.tasks;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import lombok.Getter;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StageStatusPreconditionTask implements PreconditionTask {
+
+  @Override
+  public String getPreconditionType() {
+    return "stageStatus";
+  }
+
+  @Override
+  public @Nonnull TaskResult execute(@Nonnull Stage stage) {
+    StageStatusPreconditionContext context =
+        stage.mapTo("/context", StageStatusPreconditionContext.class);
+    String stageName = context.getStageName();
+    String assertedStatus = context.getStageStatus();
+    if (stageName == null) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Stage name is required for preconditions of type %s.", getPreconditionType()));
+    }
+    if (assertedStatus == null) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Stage status is required for preconditions of type %s.", getPreconditionType()));
+    }
+    Stage foundStage =
+        stage.getExecution().getStages().stream()
+            .filter(s -> s.getName().equals(stageName))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new IllegalArgumentException(
+                        String.format(
+                            "Failed to find stage %s in execution. Please specify a valid stage name",
+                            stageName)));
+    String actualStatus = foundStage.getStatus().toString();
+    if (!actualStatus.equals(assertedStatus)) {
+      Map<String, Object> outputs =
```

Comment:
It might make sense to write a helper function to encapsulate the logic of where to put error messages.

Reply:
Agreed, will await the blessing of @robfletcher to make sure this is even the most up to date correct place to surface errors :ear:

Reply:
I think the only "centralized"/"consistent" place where errors are handled is in the `RunTaskHander` which is basically what you're replicating here. I say, just throw and let https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/RunTaskHandler.kt#L154 handle it

Reply:
@marchello2000 thank you so much, that makes sense! Updated the task and its tests accordingly :+1: